Wip heapster merge

benjaminselfridge opened the pull request:
Merge wip-heapster changes into master. This is a draft until we get the master merged back into wip-heapster.

Commit:
trying to build a CompM monad based on itrees

Commit:
marked the theorems I want to prove as Admitted

Commit:
marked "old stuff below"

Commit:
added List_def translation

Commit:
more progress on CompM_Itrees

Commit:
almost got bind_satisfies_bind to work, modulo stupid coinductive types things

Commit:
got bind_satisfies_bind working!

Commit:
changed bind_satisfies_bind to use observing equality instead of eutt

Commit:
almost got a version of bind_satisfies_bind that works, except we need to prove some sort of equivalence to replace observing that works...

Commit:
proved bind_satisfies_bind with no more admits!

Commit:
changed the definition of satisfies so that it avoids having everything satisfy spin

Commit:
whoops, removed a syntax error

Commit:
added assertM and assumingM

Commit:
commented out old Coq modules that no longer compile

Commit:
changed errorM to take a string argument

ChrisEPhifer committed:
Added `refinesFun1` to the CompM module

ChrisEPhifer committed:
Moved injectivity proof for CompM to its module

ChrisEPhifer committed:
Merge pull request #10 from GaloisInc/fix/compm-enhancements

Move some stuff from the formal-vaccine project into the CompM module

ChrisEPhifer committed:
Renamed and moved injectivity lemma to CompMExtra.v

Commit:
updated prove_refinement tactics for the new definition of errorM

Commit:
Merge branch 'wip-heapster' of github.com:GaloisInc/saw-core-coq into wip-heapster

Commit:
proved refinesM_fixM, and renamed it to refinesM_fixM_l

Commit:
proved refinesM_fixM_lr

Commit:
added the translation of bvule

Commit:
re-generated SAWCorePrelude.v from the modified Prelude.sawcore

Commit:
bug fix: binders with the same string names were causing unintended shadowing

Commit:
added a translation for the new appendString primitive

Commit:
tweaked the tacitcs for eliminating eithers in refinement

Commit:
Merge branch 'master' into wip-heapster

Commit:
updated generated Coq modules for new saw-core; FIXME: CryptolPrimitivesForSAWCore.v no longer compiles!

Commit:
fixed bvule being skipped in the translation

Commit:
re-generate prelude exports

m-yac committed:
bitvector improvements (...)
- add printing and well-behaved literals, the latter of which can be  
  simplified in a goal using compute_bvLits or prove_refinement
- remove unnecessary conversions to Nat
- add signed comparisons

m-yac committed:
implement shiftL, shiftR

m-yac committed:
export Z numeral notations from SAWCoreVectorsAsCoqVectors

m-yac committed:
add refinesM_letRecM1_fst, refinesM_maybe_l/r, assumingM rules

m-yac committed:
add SAWCoreBitvectors

m-yac committed:
prove refinesFun is a PreOrder

m-yac committed:
tweak to prove_refinement, more bitvector lemmas

m-yac committed:
add orM/andM, bitvector rewriting in hyps, and/or elim in hyps

Commit:
added escapeStringLit to escape double quotes when printing string literals

m-yac committed:
do destructing more carefully, add commented-out debug printing

m-yac committed:
make bitvector inequalities opaque

m-yac committed:
replace bvLit with intToBv, add Z literals to Coq translation

Commit:
added a special case in the translation for translating bitvectors, addressing #25

Commit:
switched to using intToBv for writing bitvector literals instead of bvLit (as the two had the same definition anyway); also changed bitvector literals to use signed integers, and switched to using the bv-sized library to do this translation instead of rolling our own; all of this in support of #25

m-yac committed:
define ProperLRTFun, prove general refinement lemmas about letRecM

m-yac committed:
make all bv functions using BITS opaque, add compute_bv_funs tactic

m-yac committed:
make changes in support of function call rewriting, add docs to tactics

m-yac committed:
prove compute_bv_funs, add refinement_proofs db, revert some changes

m-yac committed:
use eauto for intro + destructing, clean up CompMExtra a bit

m-yac committed:
add Exists and Forall ArgNames

m-yac committed:
move IntroArg stuff from refinesM to refinesFun HintDb

brianhuffman committed:
Heapster bound names (#29)

* Translate bound names so as to avoid shadowing or using reserved keywords.

* Revert "bug fix: binders with the same string names were causing unintended shadowing"

This reverts commit 2e3c63d37919939db204c52108ce7546866f14e9.

The same bug was already fixed in PR #12.

* Fix `translatePi` so that it properly translates bound variables.

This patch fixes an oversight in the #12 PR. Previously it
incorrectly failed to use the name-collision-avoiding
`translateLocalIdent` function, and it similarly failed to
update the `localEnvironment` with the bound name.

* Fix incorrect local variable environment when translating lambdas.

This patch fixes another problem that was missed in PR #12.
Previously when translating the body of a lambda, the translator
would populate the local variable environment with the *old*
variable names instead of using the new ones. This was causing
incorrect output when a lambda variable had been renamed.

* Fix incorrect local variable environment when translating terms.

This patch fixes a few more instances of redundant and/or wrong
manipulations of the local variable environment when translating
terms. Most of these changes were adapted from the wip-heapster
branch (revision 2e3c63d3).

Commit:
re-generated Coq files

Commit:
Merge branch 'master' into wip-heapster-merge

benjaminselfridge commented:
This should be ready for merge.

m-yac committed:
changes for `no_errors_incr_list`

Commit:
preparation for merge

Commit:
Merge branch 'master' into wip-heapster

Commit:
Merge branch 'wip-heapster-merge' of https://github.com/GaloisInc/saw-core-coq into wip-heapster-merge

Commit:
resolve missed conflict

Commit:
fixed pretty-printing for identifiers that start with "@" by adding parens when needed

Commit:
fixed withLocalLocalEnvironment to save the entire environment, not just parts of it

brianhuffman commented on src/Verifier/SAW/Translation/Coq/Term.hs:
```diff
@@ -289,11 +289,9 @@ asApplyAllRecognizer t = do _ <- asApp t
   result <- action
-  modify $ set localEnvironment env
-  modify $ set unavailableIdents used
+  put s
```
What about the `globalDeclarations` and `localDeclarations` fields; if `action` updates either of those, we don't want to throw away the changes, do we?

eddywestbrook replied:
Actually, propagating changes to `localDeclarations` is exactly what was causing the errors that this change fixed: if you introduce a let-binding in a local scope, its variables are no longer usable outside that local scope. As for `globalDeclarations`, I checked, and there is no place that it is modified inside a `withLocalLocalEnvironment`.

More generally, this goes back to my suggestion somewhere else that this should really be a `Reader` and not a `State` monad: `withLocalLocalEnvironment` should really just be thought of as a `local` in a reader monad.

brianhuffman replied:
My point is that the translation monad really needs to be a `Reader` + `Output` monad, and the `localDeclarations` field should be in the `Output` part, and not the `Reader` part. The `localDeclarations` field is used to keep track of all the `Constant` AST nodes that are encountered during translation, because each of them will need to be lifted out to a separate top-level Coq declaration at the end. If you treat them as part of the `Reader` state, and reset them after `withLocalLocalEnvironment`, then any `Constant` AST nodes that appear under binders will be forgotten, won't they?

brianhuffman commented on src/Language/Coq/Pretty.hs:
```diff
@@ -121,6 +121,7 @@ ppTerm p e =
     Sort s ->
       ppSort s
     Var x ->
+      parensIf (p > PrecApp && head x == '@') $
```
I'd like to see this predicate `head x == '@'` factored out into a separate function of type `Ident -> Bool`, which would probably belong in the AST module. Considering #30, we should try to minimize the code that depends on how `Ident` is implemented.

eddywestbrook replied:
Well, I'll go one step farther, and suggest that identifiers should not begin with `'@'`, and `translateIdentWithArgs` should not be prepending `'@'` to identifiers. Instead, that is just a hack to represent the notion that an identifier needs to be explicit in all its arguments, which should really be a different syntactic form in the AST. I'll implement and test this now.

eddywestbrook replied:
Ok, this is fixed with 4b2d7e09ecbbecb4bbdb296758ac962911d4ba3d

m-yac committed:
pull updated `CompM_itrees`

from https://github.com/DeepSpec/InteractionTrees/blob/7bb773007ca7ea3164f00495505888e6bf36b549/theories/ITreeSpec/CompM_ITrees.v

m-yac committed:
clean up `CompM_ITrees`

Commit:
committed latest versions of generated Coq files

Commit:
added an "explicit variable" form to the Coq AST, in place of identifiers that start with "@"

Commit:
whoops, bvToInt was actually intToBV!

Commit:
whoops, wrong definition for bvToInt

Commit:
changed the way record projections are represented in Coq, to handle #34

Commit:
Adds bvShl and enables CryptolPrimitivesForSAWCore to build

brianhuffman committed:
Merge branch 'master' into wip-heapster-merge

brianhuffman commented:
According to the CI results for GaloisInc/saw-script#1019, there are some compiler errors occurring with the ghc-8.10 builds. We'll need to address those here before merging.

brianhuffman committed:
Add GADTs language pragma to make ghc-8.10 happy.

In 8.10, GHC is pickier about requiring language extensions
to match on GADT constructors.
